apps/admin-dashboard/app/routes/_dashboard.applications.$id.accept.tsx

```diff
@@ -0,0 +1,117 @@
+import {
+  type ActionFunctionArgs,
+  json,
+  type LoaderFunctionArgs,
+  redirect,
+} from '@remix-run/node';
+import {
+  Form as RemixForm,
+  useActionData,
+  useLoaderData,
+  useNavigate,
+} from '@remix-run/react';
+
+import { Button, Form, Modal } from '@oyster/ui';
+
+import { Route } from '../shared/constants';
+import { acceptApplication, getApplication } from '../shared/core.server';
+import {
+  commitSession,
+  ensureUserAuthenticated,
+  toast,
+  user,
+} from '../shared/session.server';
+
+export async function loader({ params, request }: LoaderFunctionArgs) {
+  await ensureUserAuthenticated(request, {
+    allowAmbassador: true,
+  });
+
+  const application = await getApplication(params.id as string, [
+    'applications.firstName',
+    'applications.lastName',
+  ]);
+
+  if (!application) {
+    throw new Response(null, { status: 404 });
+  }
+
+  return json({
+    application,
+  });
+}
+
+export async function action({ params, request }: ActionFunctionArgs) {
+  const session = await ensureUserAuthenticated(request, {
+    allowAmbassador: true,
+  });
+
+  const adminId = user(session);
+
+  try {
+    await acceptApplication(params.id as string, adminId);
+
+    toast(session, {
+      message: 'Application has been accepted.',
+      type: 'success',
+    });
+
+    return redirect(Route.APPLICATIONS, {
+      headers: {
+        'Set-Cookie': await commitSession(session),
+      },
+    });
+  } catch (e) {
+    if ((e as Error).message.includes('FK_7b8cdcaabacb80df1ed8f7dbbea')) {
+      return json({
+        error: "You don't have the permissions to accept an application",
+      });
+    } else {
+      return json({
+        error: (e as Error).message,
+      });
```

Comment:
I implemented the logic so that when 'Confirm' is pressed, the function `acceptApplication` runs and we update the databases. However, I was having issues testing this because when running this, I received the following error:

> insert or update on table "applications" violates foreign key constraint "FK_7b8cdcaabacb80df1ed8f7dbbea"

After some digging it seems this means that I don't have admin access and so I can't update the database, so this limited my ability to test fully. Is there a way for me to assume admin access in the dev environment or somehow work around this?

Reply:
So actually I think this is just a bug that only applies in the development environment, and here's how to reproduce:

1. Log into the Admin Dashboard.
2. Re-run the `db:seed` script, which wipes the DB and creates a new admin record.
3. Go back into the Admin Dashboard.

What actually happens: when you go to the Admin Dashboard, you're still logged in, but the ID of the logged in user in the session is actually invalid because it no longer exists in the DB.

What should happen: when you go to the Admin Dashboard, you're deemed unauthenticated because the ID is not valid.

The reason this bug is here is because [`isUserAuthenticated`](https://github.com/colorstackorg/oyster/blob/63b72dc982e4c145ce9f40467f32793ba88a05ec/apps/admin-dashboard/app/shared/session.server.ts#L67) doesn't do a DB check for the admin. I'm hesitant to do a DB call on every route (particularly because this issue would only arise in development), but technically that would be the correct thing to do.

In the meantime, you can fix this issue by just logging out and logging back in (because it will set the correct admin ID).

Reply:
Ahh that makes sense. This fix worked 👍